src/components/Autocomplete/components/ComboBox/ComboBox.tsx

```diff
@@ -53,427 +41,374 @@ export interface ComboBoxProps {
-  ) {
-    const optionsChanged =
-      filterForOptions(navigableOptions) &&
-      nextOptions &&
-      !optionsAreEqual(navigableOptions, nextOptions);
-
-    let newNavigableOptions: (
-      | OptionDescriptor
-      | ActionListItemDescriptor
-    )[] = [];
-    if (nextActionsBefore) {
-      newNavigableOptions = newNavigableOptions.concat(nextActionsBefore);
-    }
-    if (optionsChanged || nextActionsBefore) {
-      newNavigableOptions = newNavigableOptions.concat(nextOptions);
-    }
-    if (nextActionsAfter) {
-      newNavigableOptions = newNavigableOptions.concat(nextActionsAfter);
-    }
-    newNavigableOptions = assignOptionIds(newNavigableOptions, comboBoxId);
+  const handlePopoverClose = useCallback(() => {
+    setPopoverActive(false);
+    setPopoverWasActive(false);
+  }, []);
 
-    if (optionsChanged && selectedOptions !== nextSelected) {
-      return {
-        navigableOptions: newNavigableOptions,
-        selectedOptions: nextSelected,
-      };
-    } else if (optionsChanged) {
-      return {
-        navigableOptions: newNavigableOptions,
-      };
-    } else if (selectedOptions !== nextSelected) {
-      return {selectedOptions: nextSelected};
+  const handlePopoverOpen = useCallback(() => {
+    if (!popoverActive && navigableOptions && navigableOptions.length > 0) {
+      setPopoverActive(true);
+      setPopoverWasActive(true);
     }
-    return null;
-  }
-
-  state: State = {
-    comboBoxId: this.getComboBoxId(),
-    selectedOption: undefined,
-    selectedIndex: -1,
-    selectedOptions: this.props.selected,
-    navigableOptions: [],
-    popoverActive: false,
-    popoverWasActive: false,
-  };
+  }, [navigableOptions, popoverActive]);
 
-  componentDidMount() {
-    const {options, actionsBefore, actionsAfter} = this.props;
-    const comboBoxId = this.getComboBoxId();
-    let navigableOptions: (OptionDescriptor | ActionListItemDescriptor)[] = [];
-
-    if (actionsBefore) {
-      navigableOptions = navigableOptions.concat(actionsBefore);
-    }
-    if (options) {
-      navigableOptions = navigableOptions.concat(options);
-    }
-    if (actionsAfter) {
-      navigableOptions = navigableOptions.concat(actionsAfter);
-    }
-    navigableOptions = assignOptionIds(navigableOptions, comboBoxId);
+  const visuallyUpdateSelectedOption = useCallback(
+    (
+      newOption: OptionDescriptor | ActionListItemDescriptor,
+      oldOption: OptionDescriptor | ActionListItemDescriptor | undefined,
+    ) => {
+      if (oldOption) {
+        oldOption.active = false;
+      }
+      if (newOption) {
+        newOption.active = true;
+      }
+    },
+    [],
+  );
 
-    this.setState({
-      navigableOptions,
-    });
-  }
+  const resetVisuallySelectedOptions = useCallback(() => {
+    setSelectedOption(undefined);
+    setSelectedIndex(-1);
+    navigableOptions &&
+      navigableOptions.forEach(
+        (option: OptionDescriptor | ActionListItemDescriptor) => {
```

Comment:
Typescript should be able to infer the type for the function argument based upon the shape of navigableOptions so you shouldn't need to explicitly specify it

```suggestion
        (option) => {
```